subsys/bluetooth/services/ddfs.c

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/ddfs.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_ddfs, CONFIG_BT_DDFS_LOG_LEVEL);
+
+/* DDF Application error codes */
+#define DDF_ERR_IN_PROGRESS		0x80
+#define DDF_ERR_CCC_CONFIG		0x81
+
+/* DDF Feature bits */
+#define DDF_FEAT_DM_RANGING_MODE_RTT	BIT(0)
+#define DDF_FEAT_DM_RANGING_MOCE_MCPD	BIT(1)
+
+/* DDF Opcodes */
+#define DDF_OP_DM_RANGING_MODE		0x01
+#define DDF_OP_DM_READ_CONF		0x0A
+#define DDF_OP_RESPONSE			0x10
+
+/* DDF Control Point Response Values */
+#define DDF_CP_RSP_SUCCESS		0x01
+#define DDF_CP_RSP_OP_NOT_SUPP		0x02
+#define DDF_CP_RSP_INVAL_PARAM		0x03
+#define DDF_CP_RSP_FAILED		0x04
+
+/* DFF Distance Measurement Flags */
+#define DDF_DM_RTT_PRESENT		BIT(0)
+#define DDF_DM_MCPD_PRESENT		BIT(1)
+
+/* Attribute indexes */
+#define DDF_SVC_DISTANCE_MEAS_ATTR_IDX	1
+#define DDF_SVC_AZIMUTH_MEAS_ATTR_IDX	4
+#define DDF_SVC_ELEVATION_MEAS_ATTR_IDX	7
+#define DDF_SVC_CTRL_POINT_ATTR_IDX	12
+
+static struct {
+	struct bt_gatt_indicate_params ind_params;
+	struct bt_ddfs_features dm_features;
+	const struct bt_ddfs_cb *cb;
+	bool dm_notification_enabled;
+	bool am_notification_enabled;
+	bool em_notification_enebled;
+	bool dm_ctrl_pt_indicate_enabled;
+	bool indicating;
+} ddfs_inst;
+
+struct write_ctrl_point_req {
+	uint8_t op;
+	union {
+		uint8_t ranging_mode;
+	};
+} __packed;
+
+struct ddfs_meas_notify {
+	uint8_t flags;
+	uint8_t quality;
+	uint8_t bt_addr[7];
```

Comment:
Why not just bt_addr_le_t addr Do you need some special handling or different address order here ?

Reply:
`bt_addr_le_t` and the `bt_addr_le_copy` function are not used because the address type field in the characteristic is after the address. This is different then in `bt_addr_le_t`.

Reply:
What do you think about defining struct for address here:
```
struct ddfs_bt_addr {
bt_addr_t a;
uint8_t type;
};
```

I think you could then get ride with magic numbers in loop in this source file and code would be more readable

Reply:
Thanks, I think that's a good suggestion. Modified.